### PR TITLE
Test expressions in context

### PR DESCRIPTION
**Pull request by katcharov**
JAVA-4820

### CODE REVIEW

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (katcharov) [inline driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:85]: Was reading docs `match(expr(new Document("$and"`, seems that we might add a version of expr such that the above would become:

`Filters.expr(d -> d.getInteger("x").eq(of(1))`

File: driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: Why do we need a generic parameter here?  Do we expect `DocumentExpression` to be sub-typed?

Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: There are times that users would need access to the current document as a `MapExpression`.  How would that be done?

(feel free to wait on this until after the map expression PR is merged)

Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: Right. The sub-type would allow a helper class (aka schema class, binding glass) to condense:
`doc.getDocument("c").<IntegerExpression>getArray("a").first()`
to
`doc.c.a.first()`

Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: I think this is now possible via `current().isMapOr(ofMap(parse("{}")))`. This isn't great, but (while certainly possible) it seems very unusual for documents to be maps.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: It's not unusual for a document to be a `MapExpression<Expression>`. It always is.

Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: This is true in a certain sense, but I do not think that sense is relevant. For example, technically a Java object "is" (is representable as) a `Map<String, Object>`. But to use it as such, users must use the inconvenient reflection API.

A usage seems improbable if all of us have trouble imagining semi-realistic cases where users would have to store things in collections that need to be used as if they were maps. This would mean that they are storing things that have an actual identity (expressed in the `_id` field), and then some completely arbitrary fields (that is, the programmer cannot hard-code at least some of the other fields, and none of those fields were instead represented as a map on the thing). And this is happening often enough that the isMapOr conversion is burdensome.

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:66]: Can be `static`.

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:135]: I was initially confused by `BsonDocument.parse` and thought this is a commented out code that should not be here, but then I realized that it's a relaxed JSON view of what we pass to `computed`. Could we remove the `BsonDocument.parse` part and leave only JSON?

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (stIncMale) [reply driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:85]: I am imagining that this could be something for the future.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: We discussed this and decided to go with inheritance (which would have been done in #1054), but then went back to non-inheritance due to an implementation problem that is ultimately caused by type erasure.

Added currentAsMap.

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (katcharov) [reply driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:135]: I'm not sure what to think of empty documents - it seems that we could leave them out for now. I don't think the example in the test needs to be particularly realistic, but I updated it to be a non-empty document. (I did not use ofEmptyMap because I don't expect that it applies anymore, due to non-inheritance).

File: driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:187]: [optional]

This text is different from that in https://github.com/mongodb/mongo-java-driver/pull/1059.

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (stIncMale) [reply driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java]: Yes, if we are not doing `of(new Document())`, it appears to be fine to not introduce `DocumentExpression of()`.

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:174]: Shouldn't `nested.x` be set to the max of 9 and 4 here? Based on the assertion it is set to 4 for some reason.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java
Comment (katcharov) [reply driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:174]: `of(Document.parse("{x: 9}"))` is not the current document, it is a literal. It is having its `x` value correctly set ("overwritten") to the maximum of 1 and 4.

Comment (stIncMale) [reply driver-sync/src/test/functional/com/mongodb/client/model/expressions/InContextExpressionsFunctionalTest.java:174]: Thank you, I was reading this absolutely incorrectly.

**Review by stIncMale — APPROVED** [review]

File: driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java:175]: While we don't know how will `Expression.isDocumentOr` return values of user-supplied subtypes of `DocumentExpression`, or whether it will even be possible to implement the idea of user subtypes, it at least it does not immediately seem absolutely impossible.

Returning a value of a user-defined type from a `static` method `Expressions.current()` is, on the other hand, absolutely impossible, provided that we exclude any kind of global state meddling (which we should). Thus, `Expressions.current` should return `DocumentExpression`.

**Review by stIncMale — CHANGES_REQUESTED** [review]

**Review by stIncMale — APPROVED** [review]

**Comment by katcharov** [comment]
Rebased, no conflicts

**Review by jyemin — APPROVED** [review]
LGTM

**Pull request merged**
